Core/AreaTrigger: handle AREATRIGGER_FLAG_HAS_CIRCULAR_MOVEMENT

Traesh opened the pull request:
**Changes proposed:**

-  Handle AREATRIGGER_FLAG_HAS_CIRCULAR_MOVEMENT
-  It's related to some AreaTrigger making circles around caster or position, and following caster
-  Example video : https://cdn.discordapp.com/attachments/431355792197812236/431442972639232000/2018-04-05_15-18-26.mp4

Thanks to @Seyden for the help

**Target branch(es):**

- [x] master

**Issues addressed:** Closes #


**Tests performed:**

build, tested ingame with .cast spell 193465, work perfectly (see video)


**Known issues **

None

Traesh committed:
Core/AreaTrigger Handle AREATRIGGER_FLAG_HAS_CIRCULAR_MOVEMENT

Core/AreaTrigger Fix reviews

Core/AreaTrigger Handle AreaTriggerReShape for circular movements

# Conflicts:
#	src/server/game/Entities/AreaTrigger/AreaTrigger.cpp
#	src/server/game/Entities/Object/Object.cpp
#	src/server/game/Globals/AreaTriggerDataStore.cpp

Typo

Fix reviews

Seyden commented on sql/updates/world/master/2018_04_05_01_world_at_circular.sql:
```diff
@@ -0,0 +1,17 @@
+    `InitialAngle` FLOAT NOT NULL DEFAULT 0,
+    `ZOffset` FLOAT NOT NULL DEFAULT 0,
+    `CounterClockWise` TINYINT(3) NOT NULL DEFAULT 0,
+    `CanLoop` TINYINT(3) NOT NULL DEFAULT 0,
```
maybe rename CanLoop to only Loop...

illfated replied:
too ambiguous. "LOOP" is a SQL keyword.

Seyden replied:
Ahhh shit :D didn't knew that

Seyden commented on src/server/game/Entities/AreaTrigger/AreaTriggerTemplate.h:
```diff
@@ -92,6 +92,17 @@ struct AreaTriggerScaleInfo
     } ExtraScale[MAX_AREATRIGGER_SCALE];
 };
 
+struct AreaTriggerCicularMovementInfo
```
Cicular -> Circular

Shauren commented on src/server/game/Entities/Object/Object.cpp:
```diff
@@ -644,27 +644,33 @@ void Object::BuildMovementUpdate(ByteBuffer* data, uint32 flags) const
+            *data << int32(areaTrigger->GetElapsedTimeForMovement());
+            *data << uint32(unkUint32);
+
+            *data << areaTriggerMiscTemplate->CircularMovementInfo.CircleRadius;
```
missing explicit types

Shauren commented:
also client will respond with CMSG_OBJECT_UPDATE_FAILED  (or disconnect) if some of these fields dont have valid values (will have a list of restrictions in ~1h)

Seyden commented:
@Shauren maybe you could help us with the Blend radius

Shauren commented on src/server/game/Entities/Object/Object.cpp:
```diff
@@ -644,27 +644,33 @@ void Object::BuildMovementUpdate(ByteBuffer* data, uint32 flags) const
+            uint32 unkUint32 = 1;
+
+            data->FlushBits();
+            data->WriteBit(hasAttachedCenter);
```
hasPathTarget

Seyden commented on sql/updates/world/master/2018_04_05_01_world_at_circular.sql:
```diff
@@ -0,0 +1,17 @@
+DELETE FROM `spell_areatrigger` WHERE SpellMiscId = 5016;
+INSERT INTO `spell_areatrigger` VALUES (5016, 10407, 0, 0, 0, 0, 26, 31636, 21000, 25996);
+
+INSERT INTO `spell_areatrigger_circular` VALUES (5016, 12.27271, 12.27271, 2.286397, 0, 0, 1, 25996);
```
I would remove this line.. because its not really valid..

Info: this areatrigger has to be scripted because each at spawned got a different radius and initialAngle

Traesh replied:
I'll let this as AT example if people want to test it, we'll just have to remove that when AT is scripted

illfated replied:
Codestyle-wise, it is recommended to include the column names to be populated. For the table `spell_areatrigger_circular` I guess you could keep it because it is already listed above, creating the table, but it ought to be added for the `spell_areatrigger` inserts.

joschiwald commented on src/server/game/Entities/AreaTrigger/AreaTrigger.cpp:
```diff
@@ -642,6 +656,65 @@ bool AreaTrigger::HasSplines() const
+    AreaTriggerCircularMovementInfo areaTriggerCircularMovementInfo;
+    AreaTriggerMiscTemplate const* areaTriggerMiscTemplate = GetMiscTemplate();
+
+    if (!areaTriggerMiscTemplate)
```
how is this even possible?

Traesh replied:
Possible on Ashamane because of DB spawned AreaTriggers. I can remove the check for now on TC if you prefer

joschiwald commented on src/server/game/Entities/AreaTrigger/AreaTrigger.cpp:
```diff
@@ -642,6 +656,65 @@ bool AreaTrigger::HasSplines() const
+
+void AreaTrigger::UpdateCircularMovementPosition()
+{
+    if (!GetMiscTemplate())
```
how is this even possible?

joschiwald commented on src/server/game/Entities/Object/Position.h:
```diff
@@ -130,6 +130,7 @@ struct TC_GAME_API Position
     Streamer<PackedXYZ> PositionPackedXYZStream() { return Streamer<PackedXYZ>(*this); }
     ConstStreamer<PackedXYZ> PositionPackedXYZStream() const { return ConstStreamer<PackedXYZ>(*this); }
 
+    bool IsPositionEmpty() const { return !m_positionX && !m_positionY && !m_positionZ && !m_orientation; }
```
i would prefer an optional position instead of this function

joschiwald commented:
can you provide a list of possible areatriggers that are affected?

illfated commented on src/server/game/Globals/AreaTriggerDataStore.cpp:
```diff
@@ -197,6 +199,47 @@ void AreaTriggerDataStore::LoadAreaTriggerTemplates()
+                circularMovementInfo.Float = 0.0f; \
+            }
+
+            VALIDATE_AND_SET_FLOAT(Radius,          circularMovementInfoFields[1].GetFloat());
```
https://travis-ci.org/TrinityCore/TrinityCore/builds/363028863#L1496-L1507

joschiwald commented on src/server/game/Globals/AreaTriggerDataStore.cpp:
```diff
@@ -197,6 +199,47 @@ void AreaTriggerDataStore::LoadAreaTriggerTemplates()
+
+#define VALIDATE_AND_SET_FLOAT(Float, Value) \
+            circularMovementInfo.Float = Value; \
+            if (!std::isfinite(circularMovementInfo.Float)) \
```
is it possible that we get not finite values from db?

Seyden replied:
It shouldn't be possible, but these are the checks from the client.. and as shauren said, if they're not "valid" by some reasons, there might be a client crash.. shauren wanted the check, even if its not possible.. but yeah :)

Seyden commented on src/server/game/Entities/AreaTrigger/AreaTrigger.cpp:
```diff
@@ -195,6 +205,10 @@ void AreaTrigger::Update(uint32 diff)
         if (Unit* target = GetTarget())
             GetMap()->AreaTriggerRelocation(this, target->GetPositionX(), target->GetPositionY(), target->GetPositionZ(), target->GetOrientation());
     }
+    else if (GetTemplate()->HasFlag(AREATRIGGER_FLAG_HAS_CIRCULAR_MOVEMENT))
```
We have to prevent this code

`if (GetTemplate()->HasFlag(AREATRIGGER_FLAG_HAS_ATTACHED))
    {
        if (Unit* target = GetTarget())
            GetMap()->AreaTriggerRelocation(this, target->GetPositionX(), target->GetPositionY(), target->GetPositionZ(), target->GetOrientation());
    }`

if GetTemplate()->HasFlag(AREATRIGGER_FLAG_HAS_CIRCULAR_MOVEMENT) == true

Because there are areatrigger which can have both flags..

joschiwald commented:
Yes pls until we don't have static areatriggers spawned

Seyden commented:
@joschiwald what exactly do you mean? a answer to my comment?

joschiwald committed:
some fixes

joschiwald commented on src/server/game/Entities/AreaTrigger/AreaTrigger.cpp:
```diff
@@ -642,6 +657,71 @@ bool AreaTrigger::HasSplines() const
+}
+
+AreaTriggerCircularMovementInfo AreaTrigger::GetAreaTriggerCircularMovementInfo() const
+{
```
we should save the movementinfo in the entity

joschiwald commented on src/server/game/Entities/AreaTrigger/AreaTrigger.cpp:
```diff
@@ -642,6 +657,71 @@ bool AreaTrigger::HasSplines() const
+    AreaTriggerCircularMovementInfo const& cmi = GetMiscTemplate()->CircularMovementInfo;
+
+    // AreaTrigger make exactly "TimeToTarget / Duration" loops during his life time
+    float angleDiff = 2.0f * float(M_PI) / (float(GetTimeToTarget()) / float(GetDuration())) * (float(GetTimeSinceCreated()) / float(GetDuration()));
```
it is the same like `2.0f * float(M_PI) * (float(GetTimeSinceCreated()) / float(GetTimeToTarget()))`

Seyden commented:
Bump!

Seyden committed:
Some fixes

Remove IsPositionEmpty

Try to fix Codacy Issues

pushed correct changes

remove unnecessary call on misctemplate

Seyden commented:
Maybe @Shauren does have a idea about the blend calculation.. how its calculated in the client

Seyden commented:
Just for clarification what the BlendRadius is..

![](https://j.gifs.com/86yN73.gif)

illfated commented:
Just a little F.Y.I. note regarding Codacy, in case you didn't know already.
https://github.com/purcell/sqlint/blob/master/README.md
Notice in particular this sentence:
`In time, we hope to add support for non-standard SQL variants (e.g. MySQL).`
So for the time being, you can assume that most MySQL "errors" are incorrect.

joschiwald commented on src/server/game/Entities/AreaTrigger/AreaTrigger.cpp:
```diff
@@ -158,6 +163,13 @@ bool AreaTrigger::Create(uint32 spellMiscId, Unit* caster, Unit* target, SpellIn
+    // If has circular movement but no center set, define caster as default center
+    if (GetTemplate()->HasFlag(AREATRIGGER_FLAG_HAS_CIRCULAR_MOVEMENT))
+    {
+        if (_areaTriggerCircularMovementInfo.Center.is_initialized() && _areaTriggerCircularMovementInfo.TargetGUID.is_initialized())
```
this does not match the comment below

Seyden replied:
Shit, i pushed the wrong branch.

joschiwald commented on src/server/game/Entities/AreaTrigger/AreaTrigger.cpp:
```diff
@@ -642,6 +659,60 @@ bool AreaTrigger::HasSplines() const
+        if (WorldObject* center = ObjectAccessor::GetWorldObject(*this, *_areaTriggerCircularMovementInfo.TargetGUID))
+            return center;
+
+    return &_areaTriggerCircularMovementInfo.Center->Pos;
```
crashsafe?

Seyden replied:
Same here, pushed the wrong branch

joschiwald commented on src/server/game/Entities/AreaTrigger/AreaTrigger.cpp:
```diff
@@ -642,6 +659,60 @@ bool AreaTrigger::HasSplines() const
+
+Position const* AreaTrigger::GetCircularMovementCenterPosition() const
+{
+    if (!_areaTriggerCircularMovementInfo.Center.is_initialized() && _areaTriggerCircularMovementInfo.TargetGUID.is_initialized())
```
which one takes priority? pos or guid?

Seyden replied:
Tbh, I didnt see a areatrigger which had both, so I don't think it matters

joschiwald replied:
it matters, if you want clean code

Seyden replied:
I mean there isn't any areatrigger which does have both. So why does it matter? If the areatrigger does have a targetGuid he takes that, if he doesn't he must have a center

joschiwald replied:
so, why do you check if it does not have a center pos?

Seyden replied:
You're correct.. if it doesn't matter, then the initialized check for center isn't needed at this position.

I think it was done because the setter for center and targetGuid doesn't reset the counter part. I will handle that later

joschiwald replied:
the caller should handle that

Seyden replied:
Why is that ? Should the caller manually remove targetGuid? If he wants to set center position ?

joschiwald commented on src/server/game/Entities/AreaTrigger/AreaTrigger.cpp:
```diff
@@ -642,6 +659,60 @@ bool AreaTrigger::HasSplines() const
+    angleDiff *= cmi.CounterClockwise ? 1 : -1;
+
+    // We already made one circle & can't loop
+    if (!cmi.CanLoop && (angleDiff <= 0.0f || angleDiff >= 2.0f * float(M_PI)))
```
can that work clockwise?

Seyden replied:
You mean if this calculation can work clockwise?

joschiwald replied:
yes? angleDiff is always < 0

Seyden replied:
Yeah well.. actually it doesnt even matter if clockwise or not.. if CanLoop == false.. then the whole movement doesnt work :D

joschiwald commented on src/server/game/Entities/AreaTrigger/AreaTrigger.cpp:
```diff
@@ -642,6 +659,60 @@ bool AreaTrigger::HasSplines() const
+    if (!centerPos)
+        return;
+
+    AreaTriggerCircularMovementInfo const& cmi = GetMiscTemplate()->CircularMovementInfo;
```
don't use misc template since you have it inside the entity

joschiwald commented on src/server/game/Server/Packets/AreaTriggerPackets.h:
```diff
@@ -18,10 +18,11 @@
+#include "Position.h"
 #include "ObjectGuid.h"
 #include "Optional.h"
-#include "Position.h"
```
alphabetical order

joschiwald commented on src/server/game/Entities/AreaTrigger/AreaTrigger.cpp:
```diff
@@ -642,6 +659,60 @@ bool AreaTrigger::HasSplines() const
+    if (!centerPos)
+        return;
+
+    AreaTriggerCircularMovementInfo const& cmi = _areaTriggerCircularMovementInfo;
```
i think you need to update `TimeToTarget` and `ElapsedTimeForMovement`

joschiwald commented on src/server/game/Entities/AreaTrigger/AreaTrigger.cpp:
```diff
@@ -642,6 +659,60 @@ bool AreaTrigger::HasSplines() const
+
+    AreaTriggerCircularMovementInfo const& cmi = _areaTriggerCircularMovementInfo;
+
+    // AreaTrigger make exactly "TimeToTarget / Duration" loops during his life time
```
`TimeToTarget > Duration` is that ever true?

Seyden replied:
Can look later if that's ever true

Seyden replied:
Yes its true.. look at the sql traesh provided.. there is a example.. this spell does have a duration of 21 sec but does have a timeToTarget of 31 sec

Seyden replied:
CircleCount/TimeToTarget = TimePerCircle

joschiwald replied:
wtf is CircleCount?

Seyden replied:
The Amount of Circles you want :D just for clarification.. just put a 1 in there.. TimeToTarget / Duration = CircleCount

joschiwald replied:
then the example does about 1,5 circles during its lifetime?

Seyden replied:
Exactly, these areatriggers are spawned at ymiron in maw of souls.. he spawns them with different radius variables, which results in bigger/smaller circles, which results in bigger/smaller TimeToTarget values

joschiwald replied:
then the implementation is wrong, it's Duration/TimeToTarget, what it right now?

joschiwald replied:
the implementation is Duration/TimeToTarget, the comment says TimeToTarget/Duration

Seyden replied:
You mean bigger/smaller Circles resulting in bigger/smaller TimeToTarget Values? thats only the case for Ymiron.. he does that in his script.. that has nothing do to with the system itself

Seyden replied:
Yeah the comment is wrong, i will change it

Seyden committed:
Fix reviews

Seyden committed:
Implement the last unk field (StartDelay for circular movement)

joschiwald commented on src/server/game/Entities/AreaTrigger/AreaTrigger.cpp:
```diff
@@ -642,6 +659,68 @@ bool AreaTrigger::HasSplines() const
+        return;
+
+    AreaTriggerCircularMovementInfo& cmi = _areaTriggerCircularMovementInfo;
+    if (cmi.StartDelay > GetElapsedTimeForMovement())
```
i guess the at will skip the first part of the circle path

Seyden replied:
What you mean? If i set startdelay to 5000, the areatrigger will wait 5 seconds clientside, before he starts to move

joschiwald replied:
if one circle takes 20 sec, the at starts at angle of pi/2

Seyden replied:
you're correct.. will change that

joschiwald commented on src/server/game/Entities/AreaTrigger/AreaTrigger.cpp:
```diff
@@ -642,6 +659,68 @@ bool AreaTrigger::HasSplines() const
+        angleDiff *= -1;
+
+    // We already made one circle & can't loop
+    if (!cmi.CanLoop && (angleDiff <= -circle || angleDiff >= circle))
```
std::abs can do the trick

Seyden commented on sql/updates/world/master/2018_04_05_01_world_at_circular.sql:
```diff
@@ -0,0 +1,21 @@
+    PRIMARY KEY (`SpellMiscId`)
+);
+
+DELETE FROM `spell_areatrigger` WHERE `SpellMiscId` = 5016;
```
The spell that is provided in the PR description

joschiwald commented on src/server/game/Globals/AreaTriggerDataStore.cpp:
```diff
@@ -17,11 +17,13 @@
 
 #include "AreaTriggerDataStore.h"
 #include "AreaTriggerTemplate.h"
+#include "AreaTriggerPackets.h"
```
useless include?

Seyden committed:
Fix reviews

Seyden committed:
Update comment

Seyden committed:
Fix review and implemented BlendFromRadius mechanic.. also fixed that the first movement was always at the center position.. due to GetMap()->AreaTriggerRelocation

Seyden commented:
@Shauren @joschiwald can this PR be merged? or does anyone have somthing to blame on? :D

joschiwald commented on src/server/game/Entities/AreaTrigger/AreaTrigger.h:
```diff
@@ -91,6 +92,20 @@ class TC_GAME_API AreaTrigger : public WorldObject, public GridObject<AreaTrigge
+        void ResetCircularMovementCenterPosition() { _areaTriggerCircularMovementInfo.Center.reset(); }
+
+        // Won't work if its used after the AreaTrigger was created
+        void SetCircularMovementStartDelay(uint32 delay) { _areaTriggerCircularMovementInfo.StartDelay = delay; }
```
thats why there is a todo in InitCircularMovement

Seyden replied:
There is nothing that indicates that RE_SHAPE is there for this purpose. For these specific areatriggers which i have sniffs for.. RE_SHAPE is sent immediatley after create. There is no way for me to find out what RE_SHAPE should do AND its named StartDelay for a Reason. Its only active in the first Seconds after creation. You can't delay it anymore after you delayed it.

Seyden replied:
Oh and btw.. for other circular areatriggers, there is no reshape.

Seyden commented:
@joschiwald merge it :D

Seyden commented:
I have the feeling that ElapsedTimeForMovement is a constant which is modifiable.. sniffs seems to confirm that theory

joschiwald commented on src/server/game/Entities/AreaTrigger/AreaTrigger.cpp:
```diff
@@ -642,6 +659,77 @@ bool AreaTrigger::HasSplines() const
+    float z = centerPos->GetPositionZ() + cmi.ZOffset;
+
+    // We have to use relocate here because cell update in Map->RelocateAreaTrigger is delayed
+    Relocate(x, y, z, angle);
```
explain this, Map::AreaTriggerRelocation is doing more than just a relocate

joschiwald committed:
Handle AREATRIGGER_FLAG_HAS_ATTACHED if areatrigger has circular movement

joschiwald committed:
fix build

joschiwald committed:
add function arg for future use

joschiwald committed:
Made _circularMovementInfo optional

joschiwald committed:
prepare merge
